### ISSUE TITLE
Deriving Typescript definitions directly from GitHub's OpenAPI Schema

### ISSUE THREAD

**Issue by gr2m** [octokit/types.ts#161]
The way it currently works is this:

1. We don't load the OpenAPI schema directly. Instead, we use the [github-openapi-graphql-server](https://github.com/gr2m/github-openapi-graphql-server).
2. We query for the endpoint parameter types and response schemas. 
3. `github-openapi-graphql-server` is using the de-referenced OpenAPI schema under the hood. And we don't read out the JSON Schemas for the response, instead we read out the examples and derive the types from that: https://github.com/gr2m/github-openapi-graphql-server/blob/c2cc33ce46a3ac870680cd889d106e51329b5e9f/lib/definition-to-endpoint.js#L124-L151. This has historical reasons, it's coming from a time where JSON Schemas were not available.

All this results in incomplete types, especially for reponses. See for example:

- #113
- #120
- #122
- #130
- #148
- #155
- #176
- #196
- #205

The current setup also prevents us from taking advantage of the `$ref` statements in GitHub's OpenAPI schema in order to greatly reduce duplications in the Type definitions for both endpoint parameters and responses.

Moving forward, I would do the following:

1. In `github-openapi-graphql-server`, add a way to query what version of the published OpenAPI spec was used, so that it can be loaded independently from the query response.
2. Load the JSON schema file that is not de-referenced and use it to derive all types for both, the parameters and responses.
3. Put it all together into [`scripts/update-endpoints/typescript.js`](https://github.com/octokit/types.ts/blob/5fe93f1ce3a9d94f73524dd00e0196f5c96b43bc/scripts/update-endpoints/typescript.js). 

I've yet do do a thorough research about turning JSON API / JSON Schema to TypeScript definition. I'm optimistic that there are options out there that can do the heavy lifting, but we shall see. I'm open to ideas.

I'm not yet sure how `3.` will work, exactly. I'll update this issue as soon as I have a clear picture on how it could work

**Comment by mattnotmitt** [comment]
Looks like [this project](https://github.com/ferdikoomen/openapi-typescript-codegen) is actively maintained and has done the bulk of the work already. Could be worth looking into the viability of it.

**Comment by gr2m** [comment]
Thanks! I've already started with https://github.com/drwpow/openapi-typescript. I'm working on a separate repository for types based on GItHub's OpenAPI spec here: https://github.com/gr2m/github-openapi-types/pull/1. I got pretty far already, but this is only the first step. But it will have value for folks anyway, so I'll move it to the @octokit org and publish it as `@octokit/openapi-types` once it's ready.

What we need for the SDK is slightly different, e.g. we don't differentiate parameters that are sent as URL / query / header / request body parameters. I'll be looking into this the next week
